### PR TITLE
[Distributed] refactor pynccl to support multilpe TP groups

### PR DESCRIPTION
**Pull request by youkaichao**
TL;DR:

1. The line of code change is large, because I moved code from `pynccl.py` to `pynccl_wrapper.py`
2. Before this PR, vLLM only supports one nccl communicator per process, and only support one tensor parallel group. After this PR, we can have multiple nccl communicators per process, e.g. one for tp and one for pp. Multiple tensor parallel groups are also supported.
3. Before this PR, whether we use pynccl for allreduce is changed back and forth, but actually just depends on whether we use cudagraph. So I changed it to be just one function `change_model_parallel_pynccl_allreduce` .

There are three groups in vLLM:
1. world group, created by `init_distributed_environment`
2. tp group, created by `initialize_model_parallel`
4. pp group, created by `initialize_model_parallel` (should refactor in the future, at least the name should be `init_tp_and_pp` stuff)

device communicator's life cycle should align with the corresponding group. However, currently our device communicator is actually bound to the world group, causing many troubles like init and destroy. It kind of work because we only have one tp group, but will bring trouble in future pp.

This pr fixed the problem, so that device communicators are aligned with the corresponding group.

### CODE REVIEW

**Comment by youkaichao** [comment]
There are three groups in vLLM:
1. world group, created by `init_distributed_environment`
2. tp group, created by `initialize_model_parallel`
3. pp group, created by `initialize_model_parallel` (should refactor in the future, at least the name should be `init_tp_and_pp` stuff)

device communicator's life cycle should align with the corresponding group. However, currently our device communicator is actually bound to the world group, causing many troubles like init and destroy. It kind of work because we only have one tp group, but will bring trouble in future pp.

This pr fixed the problem, so that device communicators are aligned with the corresponding group.

**Comment by youkaichao** [comment]
I feel this PR is quite large. Will break it into small pieces.

**Comment by youkaichao** [comment]
close as this is split into https://github.com/vllm-project/vllm/pull/4512 https://github.com/vllm-project/vllm/pull/4566 and https://github.com/vllm-project/vllm/pull/4591 .
